src/core/dev/allocator.cpp

```diff
@@ -252,91 +253,90 @@ xlio_registrator::~xlio_registrator()
     }
 
-    // Path for all ib contexts
     ib_context_map_t *ib_ctx_map = g_p_ib_ctx_handler_collection->get_ib_cxt_list();
-    if (likely(ib_ctx_map)) {
-        for (const auto &ib_ctx_key_val : *ib_ctx_map) {
-            p_ib_ctx_h = ib_ctx_key_val.second;
-            lkey = register_memory_single(data, size, p_ib_ctx_h, access);
-
-            if (lkey == LKEY_ERROR) {
-                deregister_memory();
-                return false;
+    uint32_t num_of_devices = ib_ctx_map->size();
+    rc = doca_mmap_set_max_num_devices(m_p_doca_mmap, num_of_devices);
+    if (DOCA_IS_ERROR(rc)) {
+        PRINT_DOCA_ERR_RET(__log_info_err, rc, "doca_mmap_set_max_num_devices");
+    }
+
+    for (const auto &ib_ctx_key_val : *ib_ctx_map) {
+        ib_ctx_handler *p_ib_ctx_h = ib_ctx_key_val.second;
+        // keep ibv mem_reg to allow doca integration with working traffic
+        uint32_t lkey = p_ib_ctx_h->mem_reg(data, size, XLIO_IBV_ACCESS_LOCAL_WRITE);
+        m_lkey_map_ib_ctx[p_ib_ctx_h] = lkey;
+
+        if ((new_dev = p_ib_ctx_h->get_doca_device())) {
+            rc = doca_mmap_add_dev(m_p_doca_mmap, new_dev);
+            if (DOCA_IS_ERROR(rc)) {
+                PRINT_DOCA_ERR_RET(__log_info_err, rc, "doca_mmap_add_dev");
             }
         }
     }
-    return true;
-}
 
-bool xlio_registrator::register_memory(void *data, size_t size, ib_ctx_handler *p_ib_ctx_h)
-{
-    return register_memory(data, size, p_ib_ctx_h, XLIO_IBV_ACCESS_LOCAL_WRITE);
+    rc = doca_mmap_set_memrange(m_p_doca_mmap, data, size);
+    if (DOCA_IS_ERROR(rc)) {
+        PRINT_DOCA_ERR_RET(__log_info_err, rc, "doca_mmap_set_memrange");
+    }
+
+    rc = doca_mmap_start(m_p_doca_mmap);
+    if (DOCA_IS_ERROR(rc)) {
+        PRINT_DOCA_ERR_RET(__log_info_err, rc, "doca_mmap_start");
+    }
+
+    return true;
 }
 
 void xlio_registrator::deregister_memory()
 {
-    ib_ctx_handler *p_ib_ctx_h;
-    ib_context_map_t *ib_ctx_map;
     uint32_t lkey;
+    ib_ctx_handler *p_ib_ctx_h;
 
-    ib_ctx_map = g_p_ib_ctx_handler_collection->get_ib_cxt_list();
-    if (ib_ctx_map) {
-        for (const auto &ib_ctx_key_val : *ib_ctx_map) {
-            p_ib_ctx_h = ib_ctx_key_val.second;
-            lkey = find_lkey_by_ib_ctx(p_ib_ctx_h);
-            if (lkey != LKEY_ERROR) {
-                p_ib_ctx_h->mem_dereg(lkey);
-                m_lkey_map_ib_ctx.erase(p_ib_ctx_h);
-            }
+    for (const auto &ib_ctx_key_val : m_lkey_map_ib_ctx) {
+        p_ib_ctx_h = ib_ctx_key_val.first;
+        lkey = find_lkey_by_ib_ctx(p_ib_ctx_h);
+        if (lkey != LKEY_ERROR) {
+            p_ib_ctx_h->mem_dereg(lkey);
         }
     }
+
     m_lkey_map_ib_ctx.clear();
+
+    if (!m_p_doca_mmap) {
+        return;
+    }
+
+    doca_error_t rc = doca_mmap_stop(m_p_doca_mmap);
+    if (DOCA_IS_ERROR(rc)) {
+        PRINT_DOCA_ERR(__log_info_err, rc, "doca_mmap_stop");
+    }
+
+    rc = doca_mmap_destroy(m_p_doca_mmap);
```

Comment:
We better set m_p_doca_mmap to null to avoid future bugs.

Reply:
Done.